06-security.md

```diff
@@ -1 +1,426 @@
+To prevent abuse an additional token must be sent with each write operation, the `XSRF-TOKEN`. The token name used in this convention is the default for AngularJS.  
+
+Typically the token is sent to the browser in the form of a cookie by the first GET request to the API, read by the JavaScript on the page and added as a request header (`X-XSRF-TOKEN`) to subsequent PUT, POST or DELETE request.  The API checks the presence of the correct header on the request. If not correct, the request fails.
+ 
+The malicious tab cannot read the contents of the `XSRF-TOKEN` cookie as it is on a different domain. 
+    
+An XSRF token is assigned to the browser on the first GET request. XSRF cookies MUST have the flag `secure` set. The API must check PUT, POST, PATCH and DELETE requests for the presence of the correct XSRF token.
+
+### Authentication
+
+Usernames and passwords are still the most commonly used way to authenticate a user especially if there is no supporting infrastructure already in place. However implementing proper authentication using usernames and passwords not a trivial task. Owasp has a [whole page](https://www.owasp.org/index.php/Authentication_Cheat_Sheet) on it. This convention describes just the minimal requirements for authenticating using a username and password. Other (important) subjects such as 'Forgotten Password' functionality or account activation is out of scope. 
+
+#### Passwords must be salted and hashed using BCrypt
+
+BCrypt is a computational heavy hashing algorithm with built in salt. If the user database is stolen if will take a lot more effort to brute force the passwords than using conventional algorithms (MD5, SHA-1, SHA-256). Note that there is a better hashing algorithm called Argon2 but this is not (yet) supported by Spring-Security.  
+
+#### A failed login attempt must always return the same response.
+
+When authentication fails the username may not exist, the password may be incorrect or the account may have been locked. If an error message is returned it should always be a generic error message ('Authentication Failed') that does not state the reason for failing. This prevents attackers from learning if actually an account exists with that username.
+
+#### Limit the number of Login attempts per unit of time.
+
+If the login form is accessible from the Internet, the number of (failed) login attempts must be limited per unit of time to prevent brute forcing. The counting must be done in the user account, not in the session (as a new session may be used for each attempt). 
+
+#### Consider using password complexity requirements and blacklists
+
+While having a these will not prevent [pattern based brute forcing techniques](https://www.youtube.com/watch?v=zUM7i8fsf0g) it will prevent users from picking too simple or common passwords.
+Good blacklists can be found [here](https://github.com/danielmiessler/SecLists/tree/master/Passwords). 
```

Comment:
Security is a cross cutting concern, so scoping is a bit arbitrary. That said, since we're doing password based authentication in most cases I thought it a good idea to include some basic guidelines.